// @azure/ai-agents/dist/commonjs/utils/utils.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolSet = exports.ToolUtility = exports.connectionToolType = void 0;
exports.isOutputOfType = isOutputOfType;
/**
 * Determines if the given output is of the specified type.
 *
 * @typeParam T - The type to check against, which extends one of the possible output parent types.
 * @param output - The action to check, which can be of type `RequiredAction`, `RequiredToolCall`, or `ToolDefinitionUnion`.
 * @param type - The type to check the action against.
 * @returns A boolean indicating whether the action is of the specified type.
 */
function isOutputOfType(output, type) {
    return output.type === type;
}
/** Types of connection tools used to configure an agent */
var connectionToolType;
(function (connectionToolType) {
    /** Bing grounding search tool */
    connectionToolType["BingGrounding"] = "bing_grounding";
    /** Microsoft Fabric tool */
    connectionToolType["MicrosoftFabric"] = "fabric_dataagent";
    /** Sharepoint tool */
    connectionToolType["SharepointGrounding"] = "sharepoint_grounding";
    /** Azure Function tool */
    connectionToolType["AzureFunction"] = "azure_function";
    /** Bing custom search tool */
    connectionToolType["BingCustomSearch"] = "bing_custom_search";
})(connectionToolType || (exports.connectionToolType = connectionToolType = {}));
const toolMap = {
    bing_grounding: "bingGrounding",
    fabric_dataagent: "fabric_dataagent",
    sharepoint_grounding: "sharepointGrounding",
    azure_function: "azureFunction",
    bing_custom_search: "bingCustomSearch",
};
/**
 * Utility class for creating various tools.
 */
class ToolUtility {
    /**
     * Creates a connection tool
     *
     * @param toolType - The type of the connection tool.
     * @param connectionIds - A list of the IDs of the connections to use.
     * @returns An object containing the definition for the connection tool
     */
    static createConnectionTool(toolType, connectionIds) {
        return {
            definition: {
                type: toolType,
                [toolMap[toolType]]: {
                    connections: connectionIds.map((connectionId) => ({ connectionId: connectionId })),
                },
            },
        };
    }
    /**
     * Creates a bing grounding search tool
     *
     * @param connectionId - The ID of the bing search connection.
     *
     * @returns An object containing the definition and resources for the bing grounding search tool
     *
     */
    static createBingGroundingTool(searchConfigurations) {
        // Ensure searchConfigurations is an array even if it's undefined
        const configs = searchConfigurations || [];
        return {
            definition: {
                type: "bing_grounding",
                bingGrounding: {
                    searchConfigurations: configs.map((searchConfiguration) => ({
                        connectionId: searchConfiguration.connectionId,
                        market: searchConfiguration?.market,
                        setLang: searchConfiguration?.setLang,
                        count: searchConfiguration?.count,
                        freshness: searchConfiguration?.freshness,
                    })),
                },
            },
        };
    }
    /**
     * Creates a file search tool
     *
     * @param vectorStoreIds - The ID of the vector store attached to this agent. There can be a maximum of 1 vector store attached to the agent.
     * @param vectorStores - The list of vector store configuration objects from Azure. This list is limited to one element. The only element of this list contains the list of azure asset IDs used by the search tool.
     * @param definitionDetails - The input definition information for a file search tool as used to configure an agent.
     *
     * @returns An object containing the definition and resources for the file search tool
     */
    static createFileSearchTool(vectorStoreIds, vectorStores, definitionDetails) {
        return {
            definition: { type: "file_search", fileSearch: definitionDetails },
            resources: { fileSearch: { vectorStoreIds: vectorStoreIds, vectorStores: vectorStores } },
        };
    }
    /**
     * Creates a code interpreter tool
     *
     * @param fileIds - A list of file IDs made available to the `code_interpreter` tool. There can be a maximum of 20 files associated with the tool.
     * @param dataSources - The data sources to be used. This option is mutually exclusive with fileIds.
     *
     * @returns An object containing the definition and resources for the code interpreter tool.
     */
    static createCodeInterpreterTool(fileIds, dataSources) {
        if (fileIds && dataSources) {
            throw new Error("Cannot specify both fileIds and dataSources");
        }
        return {
            definition: { type: "code_interpreter" },
            resources: { codeInterpreter: { fileIds: fileIds, dataSources: dataSources } },
        };
    }
    /**
     * Creates an Azure AI search tool
     *
     * @param indexConnectionId - The connection ID of the Azure AI search index.
     * @param indexName - The name of the Azure AI search index.
     *
     * @returns An object containing the definition and resources for the Azure AI search tool.
     */
    static createAzureAISearchTool(indexConnectionId, indexName, options) {
        return {
            definition: { type: "azure_ai_search" },
            resources: {
                azureAISearch: {
                    indexList: [
                        {
                            indexConnectionId: indexConnectionId,
                            indexName: indexName,
                            queryType: options?.queryType,
                            topK: options?.topK,
                            filter: options?.filter,
                        },
                    ],
                },
            },
        };
    }
    /**
     * Creates a connected agent tool
     *
     * @returns An object containing the definition for the connected agent tool.
     */
    static createConnectedAgentTool(id, name, description) {
        return {
            definition: {
                type: "connected_agent",
                connectedAgent: {
                    id: id,
                    name: name,
                    description: description,
                },
            },
        };
    }
    /**
     * Creates a function tool
     *
     * @param functionDefinition - The function definition to use.
     *
     * @returns An object containing the definition for the function tool.
     */
    static createFunctionTool(functionDefinition) {
        return {
            definition: {
                type: "function",
                function: functionDefinition,
            },
        };
    }
    /**
     * Creates an OpenApi tool
     *
     * @param openApiFunctionDefinition - The OpenApi function definition to use.
     *
     * @returns An object containing the definition for the OpenApi tool.
     */
    static createOpenApiTool(openApiFunctionDefinition) {
        return {
            definition: {
                type: "openapi",
                openapi: {
                    name: openApiFunctionDefinition.name,
                    spec: openApiFunctionDefinition.spec,
                    description: openApiFunctionDefinition.description,
                    auth: openApiFunctionDefinition.auth,
                    defaultParams: openApiFunctionDefinition.defaultParams,
                },
            },
        };
    }
}
exports.ToolUtility = ToolUtility;
/**
 * Represents a set of tools with their definitions and resources.
 */
class ToolSet {
    /** A list of tool definitions that have been added to the tool set. */
    toolDefinitions = [];
    /** A collection of resources associated with the tools in the tool set. */
    toolResources = {};
    /**
     * Adds a connection tool to the tool set.
     *
     * @param toolType - The type of the connection tool.
     * @param connectionIds - A list of the IDs of the connections to use.
     *
     * @returns An object containing the definition for the connection tool
     */
    addConnectionTool(toolType, connectionIds) {
        const tool = ToolUtility.createConnectionTool(toolType, connectionIds);
        this.toolDefinitions.push(tool.definition);
        return tool;
    }
    /**
     * Adds a file search tool to the tool set.
     *
     * @param vectorStoreIds - The ID of the vector store attached to this agent. There can be a maximum of 1 vector store attached to the agent.
     * @param vectorStores -  The list of vector store configuration objects from Azure. This list is limited to one element. The only element of this list contains the list of azure asset IDs used by the search tool.
     * @param definitionDetails - The input definition information for a file search tool as used to configure an agent.
     *
     * @returns An object containing the definition and resources for the file search tool
     */
    addFileSearchTool(vectorStoreIds, vectorStores, definitionDetails) {
        const tool = ToolUtility.createFileSearchTool(vectorStoreIds, vectorStores, definitionDetails);
        this.toolDefinitions.push(tool.definition);
        this.toolResources = { ...this.toolResources, ...tool.resources };
        return tool;
    }
    /**
     * Adds a code interpreter tool to the tool set.
     *
     * @param fileIds - A list of file IDs made available to the `code_interpreter` tool. There can be a maximum of 20 files associated with the tool.
     * @param dataSources - The data sources to be used. This option is mutually exclusive with fileIds.
     *
     * @returns An object containing the definition and resources for the code interpreter tool
     */
    addCodeInterpreterTool(fileIds, dataSources) {
        const tool = ToolUtility.createCodeInterpreterTool(fileIds, dataSources);
        this.toolDefinitions.push(tool.definition);
        this.toolResources = { ...this.toolResources, ...tool.resources };
        return tool;
    }
    /**
     * Adds an Azure AI search tool to the tool set.
     *
     * @param indexConnectionId - The connection ID of the Azure AI search index.
     * @param indexName - The name of the Azure AI search index.
     *
     * @returns An object containing the definition and resources for the Azure AI search tool
     */
    addAzureAISearchTool(indexConnectionId, indexName) {
        const tool = ToolUtility.createAzureAISearchTool(indexConnectionId, indexName);
        this.toolDefinitions.push(tool.definition);
        this.toolResources = { ...this.toolResources, ...tool.resources };
        return tool;
    }
    /**
     * Adds an OpenApi tool to the tool set.
     *
     * @param openApiFunctionDefinition - The OpenApi function definition to use.
     *
     * @returns An object containing the definition for the OpenApi tool
     */
    addOpenApiTool(openApiFunctionDefinition) {
        const tool = ToolUtility.createOpenApiTool(openApiFunctionDefinition);
        this.toolDefinitions.push(tool.definition);
        return tool;
    }
    /**
     * Adds a bing grounding search tool to the tool set.
     *
     * @param connectionId - The ID of the bing search connection.
     *
     * @returns An object containing the definition and resources for the bing grounding search tool
     */
    addBingGroundingTool(searchConfigurations) {
        const tool = ToolUtility.createBingGroundingTool(searchConfigurations);
        this.toolDefinitions.push(tool.definition);
        return tool;
    }
    /**
     * Adds a connected agent tool to the tool set.
     *
     * @param id - The ID of the connected agent.
     * @param name - The name of the connected agent.
     * @param description - The description of the connected agent.
     * @returns An object containing the definition for the connected agent tool
     */
    addConnectedAgentTool(id, name, description) {
        const tool = ToolUtility.createConnectedAgentTool(id, name, description);
        this.toolDefinitions.push(tool.definition);
        return tool;
    }
}
exports.ToolSet = ToolSet;
