// @azure/ai-projects/dist/react-native/api/telemetry/operations.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { RestError } from "@azure/core-rest-pipeline";
/**
 * Creates a new instance of the CreateTelemetryOperations class.
 * @param connectionClient - The connections client to use for operations.
 * @returns A new instance of CreateTelemetryOperations.
 */
export function createTelemetryOperations(connectionClient) {
    const state = {
        _connectionString: undefined,
    };
    return {
        /**
         * Get the Application Insights connection string associated with the Project's Application Insights resource.
         * @param options - The options parameters.
         * @returns The Application Insights connection string.
         * @throws RestError if an Application Insights connection does not exist for this project.
         */
        async getApplicationInsightsConnectionString() {
            if (!state._connectionString) {
                // Get all Application Insights connections that are marked as default
                const connections = [];
                for await (const connection of connectionClient.list({
                    connectionType: "AppInsights",
                    defaultConnection: true,
                })) {
                    connections.push(connection);
                }
                // Find the first connection that is of type "AppInsights"
                const connectionName = connections?.[0]?.name;
                if (!connectionName) {
                    throw new RestError("No Application Insights connection found.", {
                        code: "ResourceNotFound",
                        statusCode: 404,
                    });
                }
                // Get the connection with credentials
                const connection = await connectionClient.getWithCredentials(connectionName);
                if (connection.credentials?.type === "ApiKey") {
                    const apiKeyCredentials = connection.credentials;
                    if (!apiKeyCredentials.apiKey) {
                        throw new Error("Application Insights connection does not have a connection string.");
                    }
                    state._connectionString = apiKeyCredentials.apiKey;
                }
                else {
                    throw new Error("Application Insights connection does not use API Key credentials.");
                }
            }
            return state._connectionString;
        },
    };
}
